File: api/src/Sudoku/Solver.php
<?php

namespace App\Sudoku;

class Solver
{
    public function solve(Board $board): bool
    {
        if (!$this->validate($board)) {
            return false;
        }

        $rows  = array_fill(0, 9, 0);
        $cols  = array_fill(0, 9, 0);
        $boxes = array_fill(0, 9, 0);

        for ($r = 0; $r < 9; $r++) {
            for ($c = 0; $c < 9; $c++) {
                $v = $board->getValue($r, $c);
                if ($v !== 0) {
                    $bit = 1 << $v;
                    $rows[$r]  |= $bit;
                    $cols[$c]  |= $bit;
                    $boxes[intdiv($r, 3) * 3 + intdiv($c, 3)] |= $bit;
                }
            }
        }

        return $this->backtrack($board, $rows, $cols, $boxes);
    }

    public function validate(Board $board): bool
    {
        for ($row = 0; $row < 9; $row++) {
            for ($col = 0; $col < 9; $col++) {
                $num = $board->getValue($row, $col);
                if ($num === 0) continue;
                $board->setValue($row, $col, 0);
                $safe = $board->isSafe($row, $col, $num);
                $board->setValue($row, $col, $num);
                if (!$safe) return false;
            }
        }
        return true;
    }

    private function backtrack(Board $board, array &$rows, array &$cols, array &$boxes): bool
    {
        $minCount = 10;
        $bestRow  = -1;
        $bestCol  = -1;

        for ($r = 0; $r < 9; $r++) {
            for ($c = 0; $c < 9; $c++) {
                if ($board->getValue($r, $c) !== 0) continue;
                $used  = $rows[$r] | $cols[$c] | $boxes[intdiv($r, 3) * 3 + intdiv($c, 3)];
                $count = 0;
                for ($n = 1; $n <= 9; $n++) {
                    if (!($used & (1 << $n))) $count++;
                }
                if ($count === 0) return false;
                if ($count < $minCount) {
                    $minCount = $count;
                    $bestRow  = $r;
                    $bestCol  = $c;
                    if ($count === 1) break 2;
                }
            }
        }

        if ($bestRow === -1) return true;

        $box  = intdiv($bestRow, 3) * 3 + intdiv($bestCol, 3);
        $used = $rows[$bestRow] | $cols[$bestCol] | $boxes[$box];

        for ($n = 1; $n <= 9; $n++) {
            if ($used & (1 << $n)) continue;
            $bit = 1 << $n;
            $board->setValue($bestRow, $bestCol, $n);
            $rows[$bestRow]  |= $bit;
            $cols[$bestCol]  |= $bit;
            $boxes[$box]     |= $bit;

            if ($this->backtrack($board, $rows, $cols, $boxes)) return true;

            $board->setValue($bestRow, $bestCol, 0);
            $rows[$bestRow]  &= ~$bit;
            $cols[$bestCol]  &= ~$bit;
            $boxes[$box]     &= ~$bit;
        }

        return false;
    }
}

File: api_py/sudoku/solver.py
from __future__ import annotations

from .board import Board


class Solver:
    def solve(self, board: Board) -> bool:
        if not self.validate(board):
            return False

        rows = [0] * 9
        cols = [0] * 9
        boxes = [0] * 9

        for r in range(9):
            for c in range(9):
                v = board.get_value(r, c)
                if v != 0:
                    bit = 1 << v
                    rows[r] |= bit
                    cols[c] |= bit
                    boxes[(r // 3) * 3 + (c // 3)] |= bit

        return self._backtrack(board, rows, cols, boxes)

    def validate(self, board: Board) -> bool:
        for row in range(9):
            for col in range(9):
                num = board.get_value(row, col)
                if num == 0:
                    continue
                board.set_value(row, col, 0)
                safe = board.is_safe(row, col, num)
                board.set_value(row, col, num)
                if not safe:
                    return False
        return True

    def _backtrack(self, board: Board, rows: list[int], cols: list[int], boxes: list[int]) -> bool:
        min_count = 10
        best_row = -1
        best_col = -1

        found_one = False
        for r in range(9):
            for c in range(9):
                if board.get_value(r, c) != 0:
                    continue
                used = rows[r] | cols[c] | boxes[(r // 3) * 3 + (c // 3)]
                count = 0
                for n in range(1, 10):
                    if not (used & (1 << n)):
                        count += 1
                if count == 0:
                    return False
                if count < min_count:
                    min_count = count
                    best_row = r
                    best_col = c
                    if count == 1:
                        found_one = True
                        break
            if found_one:
                break

        if best_row == -1:
            return True

        box = (best_row // 3) * 3 + (best_col // 3)
        used = rows[best_row] | cols[best_col] | boxes[box]

        for n in range(1, 10):
            if used & (1 << n):
                continue
            bit = 1 << n
            board.set_value(best_row, best_col, n)
            rows[best_row] |= bit
            cols[best_col] |= bit
            boxes[box] |= bit

            if self._backtrack(board, rows, cols, boxes):
                return True

            board.set_value(best_row, best_col, 0)
            rows[best_row] &= ~bit
            cols[best_col] &= ~bit
            boxes[box] &= ~bit

        return False
